File: Data Structures & Algorithms/max-area-of-island/submission-1.go
func maxAreaOfIsland(grid [][]int) int {
    row := len(grid)
    col := len(grid[0])

    var dfs func(r, c int) int

    dfs = func(r, c int)int{
        if r < 0 || r >= row || c < 0 || c >= col || grid[r][c] != 1{
            return 0
        }

        grid[r][c] = 0

        return 1 + dfs(r, c+1) + dfs(r+1, c) + dfs(r, c-1) + dfs(r-1, c)
    }

    maxarea := 0
    for i := 0; i < row; i++{
        for j := 0; j < col; j++{
            if grid[i][j] == 1{
                cur_area := dfs(i, j)
                if cur_area > maxarea{
                    maxarea = cur_area
                }
            }
        }
    }

    return maxarea
}

File: Data Structures & Algorithms/islands-and-treasure/submission-1.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        row, col = len(grid), len(grid[0])
        q = deque()
        sett = set()

        def addroom(r, c):
            if r<0 or r>=row or c<0 or c>=col or (r,c) in sett or grid[r][c] == -1:
                return

            sett.add((r,c))
            q.append([r,c])

        # Find the Gate
        for i in range(row):
            for j in range(col):
                if grid[i][j] == 0:
                    q.append([i,j])
                    sett.add((i,j))

        dist = 0
        while q:
            for i in range(len(q)):
                r,c = q.popleft()
                grid[r][c] = dist

                addroom(r,c+1)
                addroom(r+1,c)
                addroom(r,c-1)
                addroom(r-1,c)

            dist+=1
            

        